apps/remote_control/lib/lexical/remote_control/project_node.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Lexical.RemoteControl.ProjectNode do
+  alias Lexical.RemoteControl
+  require Logger
+
+  defmodule State do
+    defstruct [:project, :paths, :cookie]
+
+    def new(project) do
+      paths = format_prepending_paths(RemoteControl.glob_paths())
+      cookie = Node.get_cookie()
+
+      struct!(__MODULE__, paths: paths, cookie: cookie, project: project)
+    end
+
+    defp format_prepending_paths(paths_as_charlists) do
+      Enum.map_join(paths_as_charlists, " -pa ", &Path.expand/1)
+    end
+  end
+
+  alias Lexical.RemoteControl.ProjectNodeSupervisor
+
+  use GenServer
+
+  def wait_until_started(project, project_listener, boot_timeout \\ 5_000) do
+    :ok = :net_kernel.monitor_nodes(true, node_type: :visible)
+
+    {:ok, node_pid} =
+      DynamicSupervisor.start_child(
+        ProjectNodeSupervisor,
+        {RemoteControl.ProjectNode, project}
+      )
+
+    node = RemoteControl.node_name(project)
+    remote_control_config = Application.get_all_env(:remote_control)
+
+    with :ok <- wait_until(boot_timeout),
+         :ok <-
+           :rpc.call(node, RemoteControl.Bootstrap, :init, [
+             project,
+             project_listener,
+             remote_control_config
+           ]) do
+      {:ok, node_pid}
+    end
+  end
+
+  defp wait_until(timeout) do
+    receive do
+      {:nodeup, _, _} ->
+        :ok
+    after
+      timeout ->
+        Logger.warn("The project node did not start after #{timeout / 1000} seconds")
+        {:error, :boot_timeout}
+    end
+  end
+
+  def start_link(project) do
+    state = State.new(project)
+    GenServer.start_link(__MODULE__, state, [])
+  end
+
+  def init(state) do
+    {:ok, state, {:continue, :start_remote_control}}
+  end
+
+  def handle_continue(:start_remote_control, state) do
+    name = RemoteControl.node_name(state.project)
```

Comment:
the responsibility of the node name really belongs this module, and not `RemoteControl`.

Reply:
Many areas use the `node_name`, should I put this in the `ProjectNode` module instead of `ProjectNode.State`?